### DISCUSSION TITLE
Use task-specific head from pretrained transformer model

### DISCUSSION THREAD

**Discussion by paulrinckens** [explosion/spaCy#8141] [Help: Coding & Implementations] [answered]
Hi!

To the best of my understanding the spacy built-in transformer component always stores the last hidden state of the underlying transformer as the tensors attribute in the  `TransformerData` object. This seems to also be the case when the used transformer model has a task-specific head for token classification (e.g. https://huggingface.co/elastic/distilbert-base-cased-finetuned-conll03-english). 

What would be the best way to get the output from a pretrained task-specific head into spaCy? 

The background to my question: I want to not train a spaCy ner component on top of the transformer component but rather use the pretrained transformer and create entities through a custom annotation_setter based on the output of the task-specific head.

Thanks!

**Comment by oadams** [comment]
I'm also interested in how to do this. If I understand you correctly, I have the same issue. I don't want to hijack, but I'll share my situation too.

I have a custom PyTorch model that feeds the final hidden token representation through a linear transformation to perform multi-label classification of each token. I can load this model in and add it to the spacy pipeline using `TransformerModel`. However, when I feed a doc through this pipeline, the `doc._.trf_data` only gives the hidden representations. It appears to be ignoring the final output that transforms the hidden representations into label space.

One option is to create another custom pipeline component that loads the final weight layer of my model and applies it to the token representations `doc._.trf_data`. But this feels bad because I'm decomposing the one PyTorch model into two pipeline components, and reloading the model.

The other thought I had involved digging into and perhaps making my own version of TransformerModel. I started looking at how things were done in [transformer_model.py](https://github.com/explosion/spacy-transformers/blob/7fd2a013b007c2a9d808bda46dd485de74e61672/spacy_transformers/layers/transformer_model.py) but then thought it would be better to ask for guidance what the best thing to do is.

My spaCy v2 approach was pretty straightforward: I had a custom pipe component that loaded the Pytorch model, took the final output of the model and modified the token attributes appropriately.

**Comment by polm** [comment, marked as answer]
Sorry for the delayed reply.

There is no special support for using Transformer heads for tasks besides getting basic sentence representations. The strategy @oadams describes of using a component to set annotations should still be possible in v3 and the most straightforward way to do things.
